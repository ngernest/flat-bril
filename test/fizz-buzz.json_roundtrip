{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v1"
          ],
          "dest": "index",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.0"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2",
            "input"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [
            "for.body.0",
            "for.end.0"
          ],
          "op": "br"
        },
        {
          "label": "for.body.0"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v6",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "dest": "div3",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "div3"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v10",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "isFizz",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v14",
          "op": "const",
          "type": "int",
          "value": 5
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "div5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "div5"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v17",
          "op": "const",
          "type": "int",
          "value": 5
        },
        {
          "args": [
            "v16",
            "v17"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v18",
            "v19"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v20"
          ],
          "dest": "isBuzz",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "isFizz"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "v22"
          ],
          "funcs": [],
          "labels": [
            "then.21",
            "else.21"
          ],
          "op": "br"
        },
        {
          "label": "then.21"
        },
        {
          "args": [
            "isBuzz"
          ],
          "dest": "v24",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "v24"
          ],
          "funcs": [],
          "labels": [
            "then.23",
            "else.23"
          ],
          "op": "br"
        },
        {
          "label": "then.23"
        },
        {
          "dest": "v25",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v26",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v25",
            "v26"
          ],
          "dest": "v27",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v27"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v28",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.23"
          ],
          "op": "jmp"
        },
        {
          "label": "else.23"
        },
        {
          "dest": "v29",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v30",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v29",
            "v30"
          ],
          "dest": "v31",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v31"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v32",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "endif.23"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.21"
          ],
          "op": "jmp"
        },
        {
          "label": "else.21"
        },
        {
          "args": [
            "isBuzz"
          ],
          "dest": "v34",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "bool"
        },
        {
          "args": [
            "v34"
          ],
          "funcs": [],
          "labels": [
            "then.33",
            "else.33"
          ],
          "op": "br"
        },
        {
          "label": "then.33"
        },
        {
          "dest": "v35",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v36",
          "op": "const",
          "type": "int",
          "value": 3
        },
        {
          "args": [
            "v35",
            "v36"
          ],
          "dest": "v37",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v37"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v38",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.33"
          ],
          "op": "jmp"
        },
        {
          "label": "else.33"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v39",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v39"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v40",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "endif.33"
        },
        {
          "label": "endif.21"
        },
        {
          "args": [
            "index"
          ],
          "dest": "v41",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v42",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v41",
            "v42"
          ],
          "dest": "v43",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v43"
          ],
          "dest": "index",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.0"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.0"
        }
      ],
      "name": "main"
    }
  ]
}
