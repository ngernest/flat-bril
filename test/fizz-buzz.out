@main
	op: const		dest: v1		type: int		value: 1	
	op: id   		dest: index		type: int		args: ["v1"]	
.for.cond.0
	op: id   		dest: v2		type: int		args: ["index"]	
	op: lt   		dest: v4		type: bool		args: ["v2", "input"]	
	op: br   		args: ["v4"]		labels: ["for.body.0", "for.end.0"]
.for.body.0
	op: id   		dest: v5		type: int		args: ["index"]	
	op: const		dest: v6		type: int		value: 3	
	op: div  		dest: v7		type: int		args: ["v5", "v6"]	
	op: id   		dest: div3		type: int		args: ["v7"]	
	op: id   		dest: v8		type: int		args: ["div3"]	
	op: const		dest: v9		type: int		value: 3	
	op: mul  		dest: v10		type: int		args: ["v8", "v9"]	
	op: id   		dest: v11		type: int		args: ["index"]	
	op: eq   		dest: v12		type: bool		args: ["v10", "v11"]	
	op: id   		dest: isFizz		type: bool		args: ["v12"]	
	op: id   		dest: v13		type: int		args: ["index"]	
	op: const		dest: v14		type: int		value: 5	
	op: div  		dest: v15		type: int		args: ["v13", "v14"]	
	op: id   		dest: div5		type: int		args: ["v15"]	
	op: id   		dest: v16		type: int		args: ["div5"]	
	op: const		dest: v17		type: int		value: 5	
	op: mul  		dest: v18		type: int		args: ["v16", "v17"]	
	op: id   		dest: v19		type: int		args: ["index"]	
	op: eq   		dest: v20		type: bool		args: ["v18", "v19"]	
	op: id   		dest: isBuzz		type: bool		args: ["v20"]	
	op: id   		dest: v22		type: bool		args: ["isFizz"]	
	op: br   		args: ["v22"]		labels: ["then.21", "else.21"]
.then.21
	op: id   		dest: v24		type: bool		args: ["isBuzz"]	
	op: br   		args: ["v24"]		labels: ["then.23", "else.23"]
.then.23
	op: const		dest: v25		type: int		value: 0	
	op: const		dest: v26		type: int		value: 1	
	op: sub  		dest: v27		type: int		args: ["v25", "v26"]	
	op: print		args: ["v27"]	
	op: const		dest: v28		type: int		value: 0	
	op: jmp  		labels: ["endif.23"]
.else.23
	op: const		dest: v29		type: int		value: 0	
	op: const		dest: v30		type: int		value: 2	
	op: sub  		dest: v31		type: int		args: ["v29", "v30"]	
	op: print		args: ["v31"]	
	op: const		dest: v32		type: int		value: 0	
.endif.23
	op: jmp  		labels: ["endif.21"]
.else.21
	op: id   		dest: v34		type: bool		args: ["isBuzz"]	
	op: br   		args: ["v34"]		labels: ["then.33", "else.33"]
.then.33
	op: const		dest: v35		type: int		value: 0	
	op: const		dest: v36		type: int		value: 3	
	op: sub  		dest: v37		type: int		args: ["v35", "v36"]	
	op: print		args: ["v37"]	
	op: const		dest: v38		type: int		value: 0	
	op: jmp  		labels: ["endif.33"]
.else.33
	op: id   		dest: v39		type: int		args: ["index"]	
	op: print		args: ["v39"]	
	op: const		dest: v40		type: int		value: 0	
.endif.33
.endif.21
	op: id   		dest: v41		type: int		args: ["index"]	
	op: const		dest: v42		type: int		value: 1	
	op: add  		dest: v43		type: int		args: ["v41", "v42"]	
	op: id   		dest: index		type: int		args: ["v43"]	
	op: jmp  		labels: ["for.cond.0"]
.for.end.0
